elfo-core/src/message/repr.rs

```diff
@@ -0,0 +1,340 @@
+
+impl MessageTypeId {
+    #[inline]
+    pub const fn new(vtable: &'static MessageVTable) -> Self {
+        Self(vtable as *const _ as *const ())
+    }
+
+    pub(super) const fn any() -> Self {
+        Self(ptr::null())
+    }
+}
+
+impl PartialEq for MessageTypeId {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        ptr::eq(self.0, other.0)
+    }
+}
+
+// === MessageRepr ===
+
+/// A message representation as a cpp-style object.
+///
+/// Initially, it's created from a typed message as [`MessageRepr<M>`], then
+/// * for [`AnyMessage`]: moved directly to heap
+/// * for [`Envelope`]: becomes a part and whole envelope is moved to heap
+///
+/// All subsequent accesses are done via `NonNull<MessageRepr>` and require
+/// using the virtual table ([`MessageVTable`]) for all operations.
+///
+/// [`AnyMessage`]: crate::AnyMessage
+/// [`Envelope`]: crate::Envelope
+// `vtable` must be first
+#[repr(C)]
+// It's `pub` only because used in private methods of `AnyMessage`.
+// Actually, it's not reexported at all.
+#[doc(hidden)]
+pub struct MessageRepr<M = ()> {
+    pub(crate) vtable: &'static MessageVTable,
+    pub(crate) data: M,
+}
+
+impl<M: Message> MessageRepr<M> {
+    /// Creates a new typed `MessageRepr` on stack.
+    /// Cannot be created for `AnyMessage` (which also implements `Message`).
+    pub(crate) fn new(message: M) -> Self {
+        debug_assert_ne!(M::_type_id(), MessageTypeId::any());
+
+        Self {
+            vtable: message._vtable(),
+            data: message,
+        }
+    }
+}
+
+// Protection against footgun.
+assert_not_impl_any!(MessageRepr: Clone);
+
+impl<M: Message> Clone for MessageRepr<M> {
+    fn clone(&self) -> Self {
+        Self {
+            vtable: self.vtable,
+            data: self.data.clone(),
+        }
+    }
+}
+
+// === MessageVTable ===
+
+/// Message Virtual Table.
+// ~
+// TODO: this struct is big enough and takes several cache lines.
+//       add `repr(C)` and reorder by frequency of access for better locality.
+// Reexported in `elfo::_priv`.
+#[doc(hidden)]
+#[non_exhaustive] // must be created only via `MessageVTable::new()`
+pub struct MessageVTable {
+    pub(super) repr_layout: alloc::Layout, // of `MessageRepr<M>`
+    pub(super) name: &'static str,
+    pub(super) protocol: &'static str,
+    pub(super) labels: [Label; 2],    // protocol + name for `metrics`
+    pub(super) dumping_allowed: bool, // TODO: introduce `DumpingMode`.
+    #[cfg(feature = "network")]
+    pub(super) read_msgpack: unsafe fn(&[u8], NonNull<MessageRepr>) -> Result<(), decode::Error>,
+    #[cfg(feature = "network")]
+    #[allow(clippy::type_complexity)]
+    pub(super) write_msgpack:
+        unsafe fn(NonNull<MessageRepr>, &mut Vec<u8>, usize) -> Result<(), encode::Error>,
+    pub(super) debug: unsafe fn(NonNull<MessageRepr>, &mut fmt::Formatter<'_>) -> fmt::Result,
+    pub(super) clone: unsafe fn(NonNull<MessageRepr>, NonNull<MessageRepr>),
```

Comment:
I recently learned that you can actually annotate names in function pointer types, like

```suggestion
    pub(super) clone: unsafe fn(src: NonNull<MessageRepr>, dest: NonNull<MessageRepr>),
```

which might be a bit more descriptive, given that most arguments here have the same types.